modules/hostmatcher/hostmatcher.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2021 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package hostmatcher
+
+import (
+	"net"
+	"path/filepath"
+	"strings"
+
+	"code.gitea.io/gitea/modules/util"
+)
+
```

Comment:
we already do have allow/block lists this is just another implementation within the same codebase :/

if for sure do it bit different ...
(https://github.com/go-gitea/gitea/blob/main/modules/migrations/migrate.go#L75-L96)

and it is planed for email too: https://github.com/go-gitea/gitea/blob/main/models/user_mail.go#L53

so I think we should only have one module to do all allow/block lists if posible

Reply:
This one is better, and can cover most cases.

Reply:
The code in `modules/migrations/migrate.go` can be refactored to use the new hostmatcher, and then we can remove a third-party dependency.

But it's difficult to use matchlist to replace hostmatcher, because we have built-in networks like "external", with hostmatcher, we only need one option for allow list (we do not need ALLOW_LOCALNETWORKS here)

Reply:
hmm we should keep in mind refactoring code ...

Reply:
Oh, maybe we can still keep both `matchlist` and `hostmatcher`, make them for different tasks.

I can find more code which is more suitable for matchlist:  https://github.com/go-gitea/gitea/blob/4e8a81780ed4ff0423e3a2ac7f75265e362ca46d/services/webhook/deliver.go#L271-L287

Reply:
Just add more thoughts (hope I am not too annoying 😊)

Why I do not use "Blocked host" here: The blocked hosts can be bypassed easily. For example: if I block "bad.com", a user can set a reverse-proxy "proxy-bad.com", then the blocked list is bypassed.

So I think in future, we can also only keep the ALLOWED_DOMAINS for the migration setting.

Reply:
well block lists are for preventing general abuse, allow list to deny your proposed scenario - we need both options so admins can configure based on there thread-model